### PR TITLE
Truncating watcher history if it is too large

### LINKED ISSUE #94745: Watcher ought to limit the size of individual .watcher_history documents
### Description

Watcher results can sometimes be enormous. We have seen at least one case of a watcher history document being sent through BulkProcessor being over 1.5 GB. This can cause cluster instability, and a 1.5 GB watch result is not ever going to be very useful. We ought to limit the size of the `WatchExecutionResult` that we put in the `WatchRecord` that is going to `.watcher_history`.

### PR DESCRIPTION
**Pull request by masseyke**
This adds a new `xpack.watcher.max.history.record.size` setting (defaulting to 10 MB), and truncates any watcher history items that exceed that size before indexing them.
Closes #94745

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/watcher/history/WatchRecord.java [hunk 1]
```diff
@@ -356,5 +418,104 @@ void innerToXContent(XContentBuilder builder, Params params) throws IOException
...
+                return TRUNCATED_RECORD_KEY;
+            }
+
+            @Override
+            public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+                builder.value(TRUNCATED_RECORD_VALUE);
+                // builder.startObject();
+                // builder.field(TRUNCATED_RECORD_KEY, TRUNCATED_RECORD_VALUE);
+                // builder.endObject();
+                return builder;
```
Comment (jbaiera) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/watcher/history/WatchRecord.java:458]: I'm assuming a string here isn't going to cause any issues mapping wise?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/watcher/history/WatchRecord.java [hunk 2]
```diff
@@ -356,5 +418,104 @@ void innerToXContent(XContentBuilder builder, Params params) throws IOException
...
+            @Override
+            public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+                builder.startObject();
+                WatcherDateTimeUtils.writeDate(
+                    Field.EXECUTION_TIME.getPreferredName(),
+                    builder,
+                    watchRecord.executionResult.executionTime()
+                );
+                builder.field(Field.EXECUTION_DURATION.getPreferredName(), watchRecord.executionResult.executionDurationMs());
+                builder.field(TRUNCATED_RECORD_KEY, TRUNCATED_RECORD_VALUE);
```
Comment (jbaiera) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/watcher/history/WatchRecord.java:515]: I'm also assuming that adding a new field to the history store doesn't need any further changes to mappings or anything?

**Review by jbaiera — APPROVED** [review]
Changes LGTM

I just had two small questions around adding new fields/document value types in the watcher indices

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/watcher/history/WatchRecord.java (hunk 1 above)
Comment (masseyke) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/watcher/history/WatchRecord.java:458]: Good question. I'll go double-check that the mapping is dynamic (I haven't run into any problems doing this, so I assume it is). It's actually not just writing the string here. Oddly, it's called from here (https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/watcher/client/WatchSourceBuilder.java#L169), and it sets the key to type(), and effectively sets the value to toXContent().
Also I'll delete those commented-out lines I apparently left in.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/watcher/history/WatchRecord.java (hunk 2 above)
Comment (masseyke) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/watcher/history/WatchRecord.java:515]: It looks like `input` doesn't specify dynamic, so it defaults to dynamic (https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/template-resources/src/main/resources/watch-history.json#L108). And `result` is explicitly dynamic (https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/template-resources/src/main/resources/watch-history.json#L136).

**Pull request merged**